src/pubsub/ua_pubsub_reader.c

```diff
@@ -1449,17 +1536,75 @@ UA_Server_DataSetReader_process(UA_Server *server, UA_DataSetReader *dataSetRead
             }
         }
     }
+
+    UA_LOG_DEBUG(&server->config.logger, UA_LOGCATEGORY_SERVER, "DataSetReader '%.*s': received a network message", 
+        (int) dataSetReader->config.name.length, dataSetReader->config.name.data);
+#ifdef UA_ENABLE_PUBSUB_MONITORING
+    /* if previous reader state was error (because we haven't received messages and ran into timeout) we should set the state back to operational */
+    if (dataSetReader->state != UA_PUBSUBSTATE_OPERATIONAL) {
+        UA_DataSetReader_setPubSubState(server, UA_PUBSUBSTATE_OPERATIONAL, dataSetReader);
```

Comment:
Why is the DSR set to OPERATIONAL implicitly?

Reply:
Good question ...
The specification only states the following:
"the DataSetReader State shall be changed to Error_3 until the next DataSetMessage is received. "
https://reference.opcfoundation.org/v104/Core/docs/Part14/6.2.8/#6.2.8.6

So I assumed that after receiving a message again the state should be set back to operational. 

According to the state machine following state changes could happen:
https://reference.opcfoundation.org/v104/Core/docs/Part14/6.2.1/

ERROR - OPERATIONAL -> Error resolved (I thought this was the use case)
ERROR - DISABLED -> I thought that this was an action from the application (manual operation)

I think ERROR - PAUSED would not apply, because it's only used if a parent changes to paused or disabled ...


I changed it to this:
```C
    /* if previous reader state was error (because we haven't received messages and ran into timeout) we should set the state back to operational */
    if (dataSetReader->state == UA_PUBSUBSTATE_ERROR) {
        UA_DataSetReader_setPubSubState(server, UA_PUBSUBSTATE_OPERATIONAL, dataSetReader);
        if (server->config.pubsubConfiguration->pubsubStateChangeCallback != 0) {
            server->config.pubsubConfiguration->pubsubStateChangeCallback(dataSetReader->identifier, UA_PUBSUBSTATE_OPERATIONAL, UA_STATUSCODE_GOOD);
        }
    }
```

But maybe I have misunderstood your question.
If you like we could talk about this tomorrow.

Reply:
Yes, thanks for your references to the standard.